engines/director/lingo/xlibs/askuser.cpp

```diff
@@ -0,0 +1,146 @@
+X      mDispose                                                            --Disposes XObject instance
+SSSS               mAsk                                                                                          --Data to display in the message box.
+*/
+
+const char *AskUser::xlibName = "AskUser";
+const char *AskUser::fileNames[] = {
+	"AskUser",
+	"ASKUSER",
+	nullptr
+};
+
+static MethodProto xlibMethods[] = {
+	{ "new",					AskUser::m_new,			 0, 0,	400 },	// D4
+	{ "ask",					AskUser::m_ask,			 3, 3,	400 },	// D4
+	{ nullptr, nullptr, 0, 0, 0 }
+};
+
+AskUserXObject::AskUserXObject(ObjectType ObjectType) :Object<AskUserXObject>("AskUser") {
+	_objType = ObjectType;
+}
+
+void AskUser::open(int type) {
+	if (type == kXObj) {
+		AskUserXObject::initMethods(xlibMethods);
+		AskUserXObject *xobj = new AskUserXObject(kXObj);
+		g_lingo->exposeXObject(xlibName, xobj);
+	} else if (type == kXtraObj) {
+		// TODO - Implement Xtra
+	}
+}
+
+void AskUser::close(int type) {
+	if (type == kXObj) {
+		AskUserXObject::cleanupMethods();
+		g_lingo->_globalvars[xlibName] = Datum();
+	} else if (type == kXtraObj) {
+		// TODO - Implement Xtra
+	}
+}
+
+void AskUser::m_new(int nargs) {
+	if (nargs != 0) {
+		warning("AskUser::m_new: expected 0 arguments");
+		g_lingo->dropStack(nargs);
+	}
+	g_lingo->push(g_lingo->_currentMe);
+}
+
+void AskUser::m_ask(int nargs) {
+	if (nargs != 3) {
+		warning("AskUser::m_ask: expected 3 arguments");
+		g_lingo->dropStack(nargs);
+		g_lingo->push(Datum("Ok"));
+		return;
+	}	
+	Datum text = g_lingo->pop();
+	Datum caption = g_lingo->pop(); // ScummVM doesn't have a title for message boxes, not used
+	Datum mbType = g_lingo->pop();
+	if (text.type != STRING) {
+		warning("AskUser::m_ask: expected text to be a string, not %s", text.type2str());
+		g_lingo->push(Datum("Ok"));
+		return;
+	}
+	if (mbType.type != STRING) {
+		warning("AskUser::m_ask: expected mbType to be a string, not %s", mbType.type2str());
+		g_lingo->push(Datum("Ok"));
+		return;
+	}
+
+	Common::U32String defaultButton;
+	Common::U32StringArray altButtons;
+
+	if (mbType.u.s->equals("YesNoCancel")) {
+		defaultButton = Common::U32String("Yes");
+		altButtons.push_back(Common::U32String("No"));
+		altButtons.push_back(Common::U32String("Cancel"));
+	} else if (mbType.u.s->equals("YesNo")) {
+		defaultButton = Common::U32String("Yes");
+		altButtons.push_back(Common::U32String("No"));
+	} else if (mbType.u.s->equals("OkCancel")) {
+		defaultButton = Common::U32String("OK");
+		altButtons.push_back(Common::U32String("Cancel"));
+	} else if (mbType.u.s->equals("Ok")) {
+		defaultButton = Common::U32String("OK");
+	} else {
+		warning("AskUser::m_ask: unhandled mbType %s, falling back to Ok", mbType.u.s->c_str());
+		defaultButton = Common::U32String("OK");
+	}
+
+	g_director->_wm->clearHandlingWidgets();
+	GUI::MessageDialog dialog(Common::U32String(text.u.s->c_str()), defaultButton, altButtons);
```

Comment:
I think that a cool thing would be using `Graphics::MacDialog` here instead of our GUI dialogs. What do you think?

Of course, that one lacks `::runModal()` method, but you could implement it there, it is just a separate event loop.

Reply:
Oh neat, we have one of those? I reckon that'd be a good for e.g. replacing b_alert also, maybe make a card for it